cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1160,88 +1162,85 @@ void writer::impl::write_chunk(table_view const &table, pq_chunked_state &state)
+                              dev_bfr,
+                              ck->ck_stat_size,
+                              cudaMemcpyDeviceToHost,
+                              stream.value()));
+            stream.synchronize();
           }
         } else {
           // copy the full data
           CUDA_TRY(cudaMemcpyAsync(host_bfr.get(),
                                    dev_bfr,
                                    ck->ck_stat_size + ck->compressed_size,
                                    cudaMemcpyDeviceToHost,
-                                   state.stream.value()));
-          state.stream.synchronize();
+                                   stream.value()));
+          stream.synchronize();
           out_sink_->host_write(host_bfr.get() + ck->ck_stat_size, ck->compressed_size);
           if (ck->ck_stat_size != 0) {
-            state.md.row_groups[global_r].columns[i].meta_data.statistics_blob.resize(
-              ck->ck_stat_size);
-            memcpy(state.md.row_groups[global_r].columns[i].meta_data.statistics_blob.data(),
+            md.row_groups[global_r].columns[i].meta_data.statistics_blob.resize(ck->ck_stat_size);
+            memcpy(md.row_groups[global_r].columns[i].meta_data.statistics_blob.data(),
                    host_bfr.get(),
                    ck->ck_stat_size);
           }
         }
-        state.md.row_groups[global_r].total_byte_size += ck->compressed_size;
-        state.md.row_groups[global_r].columns[i].meta_data.data_page_offset =
-          state.current_chunk_offset + ((ck->has_dictionary) ? ck->dictionary_size : 0);
-        state.md.row_groups[global_r].columns[i].meta_data.dictionary_page_offset =
-          (ck->has_dictionary) ? state.current_chunk_offset : 0;
-        state.md.row_groups[global_r].columns[i].meta_data.total_uncompressed_size = ck->bfr_size;
-        state.md.row_groups[global_r].columns[i].meta_data.total_compressed_size =
-          ck->compressed_size;
-        state.current_chunk_offset += ck->compressed_size;
+        md.row_groups[global_r].total_byte_size += ck->compressed_size;
+        md.row_groups[global_r].columns[i].meta_data.data_page_offset =
+          current_chunk_offset + ((ck->has_dictionary) ? ck->dictionary_size : 0);
+        md.row_groups[global_r].columns[i].meta_data.dictionary_page_offset =
+          (ck->has_dictionary) ? current_chunk_offset : 0;
+        md.row_groups[global_r].columns[i].meta_data.total_uncompressed_size = ck->bfr_size;
+        md.row_groups[global_r].columns[i].meta_data.total_compressed_size   = ck->compressed_size;
+        current_chunk_offset += ck->compressed_size;
       }
     }
   }
 }
 
-std::unique_ptr<std::vector<uint8_t>> writer::impl::write_chunked_end(
-  pq_chunked_state &state, bool return_filemetadata, const std::string &column_chunks_file_path)
+std::unique_ptr<std::vector<uint8_t>> writer::impl::close(
+  std::string const &column_chunks_file_path)
 {
+  closed = true;
```

Comment:
There doesn't seem to be a check before this to skip the few lined below in case the file was already closed. Does this mean if I call `close()` repeatedly on a file, it will write the footer multiple times to the sink?

Reply:
Yeah, it is better to skip since it is already closed.